### PR TITLE
mds: check MDSMap encoder ev for validity when decoding

### PR DESCRIPTION
**Pull request by joscollin**
Fixes: https://tracker.ceph.com/issues/44288

### CODE REVIEW

File: src/mds/MDSMap.cc [hunk 1]
```diff
@@ -610,37 +610,9 @@ void MDSMap::encode(bufferlist& bl, uint64_t features) const
   }
 
   using ceph::encode;
-  if ((features & CEPH_FEATURE_PGID64) == 0) {
```
Comment (ukernel) [inline src/mds/MDSMap.cc]: should only remove code block  if ((features & CEPH_FEATURE_PGID64) == 0) and if ((features & CEPH_FEATURE_MDSENC) == 0), don't change anything else

File: src/mds/cephfs_features.h [hunk 2]
```diff
@@ -48,6 +49,7 @@
   CEPHFS_FEATURE_NAUTILUS,              \
   CEPHFS_FEATURE_DELEG_INO,             \
   CEPHFS_FEATURE_OCTOPUS,               \
+  CEPHFS_FEATURE_MAX,                   \
```
Comment (gregsfortytwo) [inline src/mds/cephfs_features.h]: This name is not idiomatic within our codebase; a _MAX is used to indicate the lowest unused value and is not itself encoded anywhere.

**Comment by joscollin** [comment]
> I think there's some confusion here. This PR nukes old clients and changes the version numbers on the encoding, but the tracker ticket is specifically talking about the ev (extended version) value, and checking that. With this PR, we still don't look at the ev except to see if we're allowed to decode something.
> As you correctly pointed out in conversation, since the ev only covers server data, we don't want to have to worry about if clients are out of date — but we do still want to require servers be new enough to read the data they're getting. This doesn't appear to do that.
> 
> In the OSDMap, you'll see a structure like
> 
> ```
> ENCODE_START(8,7,bl)
> {
>   ENCODE_START(v,1,bl); // data for clients
>   ENCODE_FINISH(bl);
>   ENCODE_START(target_v,1,bl); // data for servers only
>   ENCODE_FINISH(bl);
> }
> ```
> 
> And it conditionally selects the target_v based on the input features, but you can then increase the compatibility value from 1 once it becomes necessary.

@gregsfortytwo  You're right this patch doesn't fix the things mentioned in the tracker, as the changes suggested by @ukernel (via IRC - dropping the old clients) seems different from what was originally mentioned in the tracker. So I think I will come back to the tracker reqs (the ev fix) once Zheng's suggestions are fixed.

@ukernel Zheng could you please review my latest push?

File: src/mds/MDSMap.cc [hunk 3]
```diff
@@ -610,37 +611,22 @@ void MDSMap::encode(bufferlist& bl, uint64_t features) const
...
 
-    int32_t m = cas_pool;
-    encode(m, bl);
-    return;
-  } else if ((features & CEPH_FEATURE_MDSENC) == 0) {
-    __u16 v = 3;
-    encode(v, bl);
+  __u8 v1, v2;
+  __u16 ev;
+  if (features & CEPHFS_FEATURE_OCTOPUS) {
```
Comment (ukernel) [inline src/mds/MDSMap.cc]: you can't use CEPHFS_FEATURE_MAX here.  features are from src/include/ceph_features.h

Comment (joscollin) [reply src/mds/MDSMap.cc]: I'm not sure which constant to be used for the version beyond octopus. I had checked ceph_features.h file earlier. But I didn't find anything related to the P release you have mentioned in the IRC. Could you please clarify @ukernel ?

**Comment by ukernel** [comment]
@gregsfortytwo 

both kernel client and libcephfs already use the ev part, need to be careful if we really want to split mdsmap to two parts.

**Comment by ukernel** [comment]
I have a idea.
- first we remove the ev part completely.  MDSMap::encode() become something like:

```
ENCODE_START(5, 4, bl);
  encode(epoch, bl);
  ...
  encode(cas_pool, bl);
  __u16 ev = 15;
  encode(ev, bl);
  ENCODE_FINISH(bl);
```

File: src/mds/MDSMap.cc (hunk 3 above)
Comment (ukernel) [reply src/mds/MDSMap.cc]: feature bit for P release hasn't defined.

**Comment by joscollin** [comment]
@ukernel I have made the changes as you suggested and pushed the code. But vstart hangs at `fs volume create a` now. The 'make check' also fails `41210: error: buffer::malformed_input: void MDSMap::decode(ceph::buffer::v15_2_0::list::const_iterator&) decode past end of struct encoding`.

I've been debugging this. Any idea what went wrong?

File: src/mds/MDSMap.cc [hunk 4]
```diff
@@ -734,123 +680,141 @@ void MDSMap::sanitize(const std::function<bool(int64_t pool)>& pool_exists)
...
-      enabled = true;
+    if (ev >= 9) {
+      decode(damaged, p);
+    }
+
+    if (!decode_ev)
+      return;
+
+    // ev part
+    DECODE_START(1, p);
```
Comment (ukernel) [inline src/mds/MDSMap.cc]: you can't add DECODE_START here.

Comment (ukernel) [reply src/mds/MDSMap.cc]: add
```
if (devode_ev) {
DECODE_START
...
DECODE_END
}
``` 
at very end of this function

Comment (joscollin) [reply src/mds/MDSMap.cc]: That's what [this line](https://github.com/ceph/ceph/pull/33835/files#diff-c4d6a907104e07425dc624f7a66f572aR756) does, right?

Comment (joscollin) [reply src/mds/MDSMap.cc]: Please clarify

Comment (ukernel) [reply src/mds/MDSMap.cc]: no. don't nest the new DECODE_START/DECODE_END in old DECODE_START/DECODE_END

Comment (ukernel) [reply src/mds/MDSMap.cc]: no. don't nest the new DECODE_START/DECODE_END in old DECODE_START/DECODE_END

Comment (joscollin) [reply src/mds/MDSMap.cc]: It should be in a different scope `{}` then. I tried that with the below error.

```
/home/jcollin/workspace/ceph/src/mds/MDSMap.cc: In member function ‘void MDSMap::decode(ceph::buffer::v15_2_0::list::const_iterator&)’:
/home/jcollin/workspace/ceph/src/include/encoding.h:1358:8: error: redeclaration of ‘__u8 struct_v’
 1358 |   __u8 struct_v, struct_compat;      \
      |        ^~~~~~~~
/home/jcollin/workspace/ceph/src/mds/MDSMap.cc:767:5: note: in expansion of macro ‘DECODE_START’
  767 |     DECODE_START(1, p);
      |     ^~~~~~~~~~~~
/home/jcollin/workspace/ceph/src/include/encoding.h:1375:8: note: ‘__u8 struct_v’ previously declared here
 1375 |   __u8 struct_v;       \
      |        ^~~~~~~~
/home/jcollin/workspace/ceph/src/include/encoding.h:1457:3: note: in expansion of macro ‘__DECODE_START_LEGACY_COMPAT_LEN’
 1457 |   __DECODE_START_LEGACY_COMPAT_LEN(v, compatv, lenv, 1u, bl)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jcollin/workspace/ceph/src/mds/MDSMap.cc:690:3: note: in expansion of macro ‘DECODE_START_LEGACY_COMPAT_LEN_16’
  690 |   DECODE_START_LEGACY_COMPAT_LEN_16(5, 4, 4, p);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jcollin/workspace/ceph/src/include/encoding.h:1368:12: error: redeclaration of ‘unsigned int struct_end’
 1368 |   unsigned struct_end = bl.get_off() + struct_len;   \

```
After putting it in another scope, vstart still hangs at `fs volume create a`.

Comment (ukernel) [reply src/mds/MDSMap.cc]: push the code, so I can check it

Comment (joscollin) [reply src/mds/MDSMap.cc]: @ukernel pushed

File: src/mds/MDSMap.cc [hunk 5]
```diff
@@ -663,69 +634,47 @@ void MDSMap::encode(bufferlist& bl, uint64_t features) const
...
     encode(created, bl);
     encode(modified, bl);
-    encode(tableserver, bl);
+    encode((int32_t)0, bl);
     encode(in, bl);
     encode(inc, bl);
     encode(up, bl);
     encode(failed, bl);
     encode(stopped, bl);
-    encode(last_failure_osd_epoch, bl);
```
Comment (ukernel) [inline src/mds/MDSMap.cc]: shouldn't remove last_failure_osd_epoch. If it belongs to ev part, encode a 0 here.  besides, when ev is 5.  encode up to  last_failure_osd_epoch

File: src/mds/MDSMap.cc [hunk 6]
```diff
@@ -624,43 +624,62 @@ void MDSMap::encode(bufferlist& bl, uint64_t features) const
...
-  encode(inline_data_enabled, bl);
-  encode(enabled, bl);
-  encode(fs_name, bl);
-  encode(damaged, bl);
-  encode(balancer, bl);
-  encode(standby_count_wanted, bl);
-  encode(old_max_mds, bl);
-  encode(min_compat_client, bl);
+  ENCODE_START(6, 4, bl);
+  {
```
Comment (ukernel) [inline src/mds/MDSMap.cc:628]: the extra bracket is no needed

File: src/mds/MDSMap.cc [hunk 7]
```diff
@@ -624,43 +624,62 @@ void MDSMap::encode(bufferlist& bl, uint64_t features) const
...
+    encode((int64_t)0, bl);
+    encode(created, bl);
+    encode(modified, bl);
+    encode((int32_t)0, bl);
+    encode(in, bl);
+    encode(inc, bl);
+    encode(up, bl);
+    encode(failed, bl);
+    encode(stopped, bl);
+    encode(last_failure_osd_epoch, bl);
```
Comment (ukernel) [inline src/mds/MDSMap.cc:656]: remove following encode(...), they will not be decoded because ev is 5

File: src/mds/MDSMap.cc [hunk 8]
```diff
@@ -624,43 +624,62 @@ void MDSMap::encode(bufferlist& bl, uint64_t features) const
...
+    encode((int32_t)0, bl);
+    encode((int32_t)0, bl);
+    encode((int32_t)0, bl);
+    encode((int32_t)0, bl);
+  }
   ENCODE_FINISH(bl);
+
+  {
+    // ev part
+    ENCODE_START(1, 1, bl);
```
Comment (ukernel) [inline src/mds/MDSMap.cc:672]: you should fre-add fields that are removed above to here

File: src/mds/MDSMap.cc [hunk 9]
```diff
@@ -702,30 +637,49 @@ void MDSMap::encode(bufferlist& bl, uint64_t features) const
...
-  encode(min_compat_client, bl);
+  encode((uint8_t)0, bl);
+  encode((uint8_t)0, bl);
+  encode((bool)0, bl);
+  encode((bool)0, bl);
+  encode((uint32_t)0, bl);
+  encode((uint32_t)0, bl);
+  encode((uint32_t)0, bl);
+  encode((uint32_t)0, bl);
+  encode((uint8_t)0, bl);
```
Comment (ukernel) [inline src/mds/MDSMap.cc:664]: don't need to encode there 0's, just remove them

Comment (ukernel) [reply src/mds/MDSMap.cc:664]: or set ev to 9, encode up to damaged.

File: src/mds/MDSMap.cc [hunk 10]
```diff
@@ -777,94 +735,45 @@ void MDSMap::decode(bufferlist::const_iterator& p)
...
-    int8_t r;
-    decode(r, p);
-    if (r < 0) {
-      min_compat_client = ceph_release_t::unknown;
-    } else {
-      min_compat_client = ceph_release_t{static_cast<uint8_t>(r)};
-    }
-  } else if (ev > 14) {
-    decode(min_compat_client, p);
-  }
```
Comment (ukernel) [inline src/mds/MDSMap.cc:865]: you need to keep decode code for ev in range 6 ~ 14

Comment (joscollin) [reply src/mds/MDSMap.cc:865]: @ukernel I reverted the range 6-14. Please check. I'm doubtful if the decode ev part is correct. It looks like there are 2 ev parts now.

File: src/mds/MDSMap.cc [hunk 11]
```diff
@@ -702,30 +637,40 @@ void MDSMap::encode(bufferlist& bl, uint64_t features) const
...
+    encode(tableserver, bl);
+    encode(ever_allowed_features, bl);
+    encode(explicitly_allowed_features, bl);
+    encode(inline_data_enabled, bl);
+    encode(enabled, bl);
+    encode(damaged, bl);
+    encode(balancer, bl);
+    encode(standby_count_wanted, bl);
+    encode(old_max_mds, bl);
+    encode(min_compat_client, bl);
```
Comment (ukernel) [inline src/mds/MDSMap.cc:671]: inline_data_enabled, enabled, damaged should belong to normal part.  the ev part encode up to damaged

File: src/mds/MDSMap.cc [hunk 12]
```diff
@@ -702,30 +637,42 @@ void MDSMap::encode(bufferlist& bl, uint64_t features) const
...
   encode(failed, bl);
   encode(stopped, bl);
   encode(last_failure_osd_epoch, bl);
-  encode(ever_allowed_features, bl);
-  encode(explicitly_allowed_features, bl);
+  encode((uint8_t)0, bl);
+  encode((uint8_t)0, bl);
   encode(inline_data_enabled, bl);
   encode(enabled, bl);
-  encode(fs_name, bl);
```
Comment (ukernel) [inline src/mds/MDSMap.cc:722]: don't remove fs_name

Comment (joscollin) [reply src/mds/MDSMap.cc:722]: @ukernel fixed

File: src/mds/MDSMap.cc [hunk 13]
```diff
@@ -702,30 +637,43 @@ void MDSMap::encode(bufferlist& bl, uint64_t features) const
   encode(data_pools, bl);
   encode(cas_pool, bl);
 
-  __u16 ev = 15;
+  __u16 ev = 5;
```
Comment (ukernel) [inline src/mds/MDSMap.cc:640]: ev = 9

File: src/mds/MDSMap.cc [hunk 14]
```diff
@@ -780,26 +731,21 @@ void MDSMap::decode(bufferlist::const_iterator& p)
   __u16 ev = 1;
   if (struct_v >= 2)
     decode(ev, p);
+  if (decode_ev)
+    ceph_assert(ev == 5);
```
Comment (ukernel) [inline src/mds/MDSMap.cc:735]: ev == 9

File: src/mds/MDSMap.cc [hunk 15]
```diff
@@ -780,26 +731,21 @@ void MDSMap::decode(bufferlist::const_iterator& p)
...
-    __u32 n;
-    decode(n, p);
-    metadata_pool = n;
-  } else {
-    decode(metadata_pool, p);
-  }
+
   decode(created, p);
   decode(modified, p);
-  decode(tableserver, p);
```
Comment (ukernel) [inline src/mds/MDSMap.cc:796]: can't remove this

File: src/mds/MDSMap.cc [hunk 16]
```diff
@@ -780,26 +731,21 @@ void MDSMap::decode(bufferlist::const_iterator& p)
...
   decode(in, p);
   decode(inc, p);
   decode(up, p);
   decode(failed, p);
   decode(stopped, p);
-  if (ev >= 4)
+  if (ev >= 5)
```
Comment (ukernel) [inline src/mds/MDSMap.cc:748]: don't change this

File: src/mds/MDSMap.cc (hunk 15 above)
Comment (joscollin) [reply src/mds/MDSMap.cc:796]: Fixed this. What about `metadata_pool` then?

**Comment by joscollin** [comment]
@ukernel In addition to your suggestions, [this](https://github.com/ceph/ceph/pull/33835/commits/eede57a83518fed6d363704e7b3a91e62912ee7c) change is required to fix the vstart hangs at `fs volume create a` issue. Now the `make check` succeeds.

File: src/mds/MDSMap.cc [hunk 17]
```diff
@@ -702,30 +637,40 @@ void MDSMap::encode(bufferlist& bl, uint64_t features) const
   encode(data_pools, bl);
   encode(cas_pool, bl);
 
-  __u16 ev = 15;
+  __u16 ev = 9;
   encode(ev, bl);
   encode(compat, bl);
```
Comment (ukernel) [inline src/mds/MDSMap.cc:642]: compat should belong to ev part (encode a empty compat set here

Comment (joscollin) [reply src/mds/MDSMap.cc]: Fixed

**Comment by ukernel** [comment]
@gregsfortytwo  pls take a look

File: src/mds/MDSMap.cc [hunk 18]
```diff
@@ -865,6 +871,24 @@ void MDSMap::decode(bufferlist::const_iterator& p)
   }
 
   DECODE_FINISH(p);
+
+  {
```
Comment (ukernel) [inline src/mds/MDSMap.cc:875]: remove this unneeded backet.  Otherwise looks good

**Comment by joscollin** [comment]
> This looks good in broad strokes! :)
> 
> But you should
> 
>     * squash into one commit
> 
>     * make a comment in the commit about dropping the encoding for old clients (I assume they are already blocked from connecting for other reasons, right?)
> 
> 
> And finally, a mon/MDS running with this code can't share working MDSMaps with an Octopus MDS, right? So we need to make sure that doesn't happen (I assume by changing the MDS protocol number one last time, but AFAIK that hasn't happened yet since Octopus branched and this PR definitely needs it), right?

@gregsfortytwo Fixed the comments. Please check.

**Comment by ukernel** [comment]
Few more things
- the commit message does not reflect the changes.
- with this change. it's better to let monitor only encode full mdsmap for mds, skip the new extented part for clients

**Comment by joscollin** [comment]
>     * with this change. it's better to let monitor only encode full mdsmap for mds, skip the new extented part for clients

@ukernel 
I think you meant the entire ev part (everything below `encode(ev, bl);`)? Putting the new ev part alone in a condition fails the test during make check.

**Comment by ukernel** [comment]
the title should be something like: update mdsmap encoding to separate common fields and server-only fields.

File: src/mds/MDSMap.cc (hunk 10 above)
Comment (ukernel) [reply src/mds/MDSMap.cc:865]: > > ```
> > * with this change. it's better to let monitor only encode full mdsmap for mds, skip the new extented part for clients
> > ```
> 
> @ukernel
> I think you meant the entire ev part (everything below `encode(ev, bl);`)? Putting the new ev part alone in a condition fails the test during make check.

client does not need fields from 'compat to min_compat_client'. it's better to add a parameter to MDSMap message 's constructor. if message is for client, skip encoding them

Comment (joscollin) [reply src/mds/MDSMap.cc:865]: @ukernel I have pushed a sample of the client checking for now. We may need to expand it.

I'm not so happy with the fix as the MDSMap object can't be modified to update `is_client`. Could you please take a look?

File: src/mds/MDSMap.cc [hunk 19]
```diff
@@ -616,79 +616,88 @@ std::string MDSMap::mds_info_t::human_name() const
...
+  encode(max_file_size, bl);
+  encode(max_mds, bl);
+  encode(mds_info, bl, features);
+  encode(data_pools, bl);
+  encode(cas_pool, bl);
 
-    int32_t m = cas_pool;
-    encode(m, bl);
+  if (epoch == 0) {
+    ENCODE_FINISH(bl);
```
Comment (ukernel) [inline src/mds/MDSMap.cc]: why do you add this

File: src/mds/MDSMap.cc [hunk 20]
```diff
@@ -643,10 +643,8 @@ void MDSMap::encode(bufferlist& bl, uint64_t features) const
   encode(data_pools, bl);
   encode(cas_pool, bl);
 
-  if (epoch == 0) {
-    ENCODE_FINISH(bl);
+  if (is_client)
```
Comment (ukernel) [inline src/mds/MDSMap.cc:646]: not stop here. I mean skip fields from 'compat' to 'min_compat_client'

File: src/mds/MDSMap.h [hunk 21]
```diff
@@ -543,6 +543,7 @@ class MDSMap {
   static bool state_transition_valid(DaemonState prev, DaemonState next);
 
   CompatSet compat;
+  bool is_client = false;
```
Comment (ukernel) [inline src/mds/MDSMap.h:546]: you should add a field to MMDSMap instead here. also add a parameter (with default value) to MDSMap::encode

File: src/mds/MDSMap.cc (hunk 19 above)
Comment (joscollin) [reply src/mds/MDSMap.cc]: This line is the older code. It's not there in MDSMap.cc now

File: src/mds/MDSMap.cc (hunk 20 above)
Comment (joscollin) [reply src/mds/MDSMap.cc:649]: @ukernel I'm confused with this comment as there are multiple compat fields. Do you mean skip the new ev part from 'compat' to 'min_compat_client' ? The below one?

```
    // ev part                                                                                                                                                                                                                                                                  
    ENCODE_START(1, 1, bl);                                                                                                                                                                                                                                                     
    encode(compat, bl);                                                                                                                                                                                                                                                         
    encode(metadata_pool, bl);                                                                                                                                                                                                                                                  
    encode(tableserver, bl);                                                                                                                                                                                                                                                    
    encode(ever_allowed_features, bl);                                                                                                                                                                                                                                          
    encode(explicitly_allowed_features, bl);                                                                                                                                                                                                                                    
    encode(balancer, bl);                                                                                                                                                                                                                                                       
    encode(standby_count_wanted, bl);                                                                                                                                                                                                                                           
    encode(old_max_mds, bl);                                                                                                                                                                                                                                                    
    encode(min_compat_client, bl); 
    ENCODE_FINISH(bl);                                                                                                                                                                                                                                                          
```

Comment (ukernel) [reply src/mds/MDSMap.cc:649]: yes

Comment (joscollin) [reply src/mds/MDSMap.cc:649]: Okay, so in this particular case of client, there would be 0 encodings in the old ev part, without the corresponding field in the new ev part. Is that correct?

```
  encode((int64_t)0, bl);                                                                                                                                                                                                                                                       
  encode(created, bl);                                                                                                                                                                                                                                                          
  encode(modified, bl);                                                                                                                                                                                                                                                         
  encode((int32_t)0, bl);
```

Comment (ukernel) [reply src/mds/MDSMap.cc:649]: no. you should update MDSMap::decode instead here

File: src/mds/MDSMap.cc [hunk 22]
```diff
@@ -614,81 +614,87 @@ std::string MDSMap::mds_info_t::human_name() const
...
+  encode(failed, bl);
+  encode(stopped, bl);
+  encode(last_failure_osd_epoch, bl);
+  encode((uint8_t)0, bl);
+  encode((uint8_t)0, bl);
+  encode(inline_data_enabled, bl);
+  encode(enabled, bl);
+  encode(fs_name, bl);
+  encode(damaged, bl);
+  encode(is_client, bl);
```
Comment (ukernel) [inline src/mds/MDSMap.cc:668]: it's better to rename this to encode_ev

**Comment by joscollin** [comment]
Jenkins test make check

**Comment by joscollin** [comment]
Jenkins test make check

File: src/mds/MDSMap.cc [hunk 23]
```diff
@@ -748,9 +754,11 @@ void MDSMap::sanitize(const std::function<bool(int64_t pool)>& pool_exists)
 void MDSMap::decode(bufferlist::const_iterator& p)
 {
   std::map<mds_rank_t,int32_t> inc;  // Legacy field, parse and drop
-
   cached_up_features = 0;
-  DECODE_START_LEGACY_COMPAT_LEN_16(5, 4, 4, p);
+  bool decode_ev;
+
+  DECODE_START_LEGACY_COMPAT_LEN_16(6, 4, 4, p);
+  decode_ev = struct_v >= 6;
```
Comment (ukernel) [inline src/mds/MDSMap.cc:761]: remove this line

File: src/mds/MDSMap.cc [hunk 24]
```diff
@@ -863,8 +870,24 @@ void MDSMap::decode(bufferlist::const_iterator& p)
   } else if (ev > 14) {
     decode(min_compat_client, p);
   }
-
+  if (decode_ev)
+    decode(encode_ev, p);
```
Comment (ukernel) [inline src/mds/MDSMap.cc:874]: if (struct_v >= 6)
      decode(decode_ev, p);
else
      decode_ev = false;

Comment (joscollin) [reply src/mds/MDSMap.cc:874]: @ukernel We are using  `decode_ev` before that - just below `// kclient ignores everything from here` for doing a `ceph_assert`

Comment (joscollin) [reply src/mds/MDSMap.cc:875]: @ukernel Done, as per your irc reply

File: src/mds/MDSMap.cc [hunk 25]
```diff
@@ -863,8 +871,24 @@ void MDSMap::decode(bufferlist::const_iterator& p)
...
   DECODE_FINISH(p);
+
+  // ev part
+  if (decode_ev && !encode_ev) {
```
Comment (ukernel) [inline src/mds/MDSMap.cc:879]: this logic is reverse

Comment (joscollin) [reply src/mds/MDSMap.cc]: @ukernel Fixed. Happened because of find/replace of the previous variable name `is_client`.

File: src/mds/MDSMap.cc [hunk 26]
```diff
@@ -863,8 +871,24 @@ void MDSMap::decode(bufferlist::const_iterator& p)
   } else if (ev > 14) {
     decode(min_compat_client, p);
   }
-
+  if (decode_ev)
+    decode(encode_ev, p);
   DECODE_FINISH(p);
+
+  // ev part
+  if (decode_ev && encode_ev) {
```
Comment (ukernel) [inline src/mds/MDSMap.cc:879]: use single variable 'decode_ev' is enough

File: src/mds/MDSMap.cc (hunk 24 above)
Comment (ukernel) [reply src/mds/MDSMap.cc:875]: this code is less clear than I suggested

File: src/mds/MDSMap.h [hunk 27]
```diff
@@ -546,6 +547,7 @@ class MDSMap {
   // base map
   epoch_t epoch = 0;
   bool enabled = false;
+  bool encode_ev = true;
```
Comment (ukernel) [inline src/mds/MDSMap.h:550]: is this used ?

Comment (joscollin) [reply src/mds/MDSMap.h:550]: at the moment - No.

**Comment by joscollin** [comment]
jenkins test dashboard backend

File: src/mds/MDSMap.cc [hunk 28]
```diff
@@ -780,17 +787,16 @@ void MDSMap::decode(bufferlist::const_iterator& p)
   __u16 ev = 1;
   if (struct_v >= 2)
     decode(ev, p);
+
+  if (struct_v >= 6)
+    ceph_assert(ev == 9);
```
Comment (batrick) [inline src/mds/MDSMap.cc:792]: add comment explanation

File: src/mds/MDSMap.cc [hunk 29]
```diff
@@ -614,81 +614,87 @@ std::string MDSMap::mds_info_t::human_name() const
...
-    }
-    encode(data_pools, bl);
-    encode(cas_pool, bl);
+  __u16 ev = 9;
+  encode(ev, bl);
+  {
+    CompatSet _compat;
+    encode(_compat, bl);
+  }
+  encode((int64_t)0, bl);
```
Comment (batrick) [inline src/mds/MDSMap.cc:652]: What does this correspond to?

File: src/mds/MDSMap.cc [hunk 30]
```diff
@@ -614,81 +614,87 @@ std::string MDSMap::mds_info_t::human_name() const
...
+  encode(created, bl);
+  encode(modified, bl);
+  encode((int32_t)0, bl);
```
Comment (batrick) [inline src/mds/MDSMap.cc:655]: ditto

File: src/mds/MDSMap.cc (hunk 29 above)
Comment (joscollin) [reply src/mds/MDSMap.cc:652]: metadata_pool. See https://github.com/ceph/ceph/pull/33835/files#diff-c4d6a907104e07425dc624f7a66f572aR676

File: src/mds/MDSMap.cc (hunk 30 above)
Comment (joscollin) [reply src/mds/MDSMap.cc:655]: tableserver. See https://github.com/ceph/ceph/pull/33835/files#diff-c4d6a907104e07425dc624f7a66f572aR677

**Comment by joscollin** [comment]
retest this please

**Comment by joscollin** [comment]
jenkins test make check

**Comment by joscollin** [comment]
jenkins test make check

**Comment by joscollin** [comment]
@batrick @gregsfortytwo ping

**Comment by gregsfortytwo** [comment]
Hmm @batrick perhaps?
I think we want to end up where the OSDMap is, having a client-intelligible portion that everything reads and a server-only bit that is versioned independently at the end, and carefully choosing which data ends up where. A new MDSMap class is one way to do that, but I'm not sure it's materially different than just eg encode_legacy() being called from encode() when there aren't enough features present.

The way this is organized now is pretty weird, though. Given that it's gating on SERVER_PACIFIC, why do we bother keeping it decodable by older versions? I'm having trouble identifying if that's *correct* with the way it's all rearranged. Certainly making a clean break and having the old encode/decode and the new shiny easier-to-partition one seems sensible to avoid having those kinds of issues.

**Comment by batrick** [comment]
We're going in a different direction: https://github.com/ceph/ceph/pull/33835#pullrequestreview-439603140

**Pull request closed without merging**